Add x86-64 assembly two-fer mentoring notes

bergjohan opened the pull request.

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
```diff
@@ -0,0 +1,93 @@
+
+### Common errors
+
+- Forgetting to save registers with callee-save status (`rbx`, `rbp`, `r12-r15`).
```
Perhaps we can show _why_ this is erroneous? What happens when you forget that? Does it still pass the tests? Why is it problematic?

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
```diff
@@ -0,0 +1,93 @@
+
+### Common suggestions
+
+- Move the string copy code into its own function or macro to avoid code duplication.
```
I love these suggestions. It might be wise to note write them as "copyable" content and more like:

- If a student has code duplication, help them find a way to de-duplicate it, for example by having multiple functions or show them that macros exist and how they work.

We probably usually don't want to "give a way the solution".

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
```diff
@@ -0,0 +1,93 @@
+
+- Move the string copy code into its own function or macro to avoid code duplication.
+
+- Use registers instead of memory for saving values as they are faster.
```
Great suggestion. Many student would like to know _why_ this is. I recommend providing at least one good article that shows this, but more is nice too.

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
```diff
@@ -0,0 +1,93 @@
+
+- Prefer `lea rax, [rel symbol]` to `mov rax, symbol`.
+
+    > There is absolutely no reason to use absolute addresses for simple memory operands. Rip-relative addresses make instructions shorter, they eliminate the need for relocation at load time, and they are safe to use in all systems. ([https://www.agner.org/optimize/optimizing_assembly.pdf](https://www.agner.org/optimize/optimizing_assembly.pdf), p. 21)
```
Great!! Exactly what I mean ❤️

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
```diff
@@ -0,0 +1,93 @@
+
+- Clearing the direction flag using `cld` is redundant.
+
+    > The direction flag `DF` in the `%rFLAGS` register must be clear (set to “forward” direction) on function entry and return. ([https://www.uclibc.org/docs/psABI-x86_64.pdf](https://www.uclibc.org/docs/psABI-x86_64.pdf), p. 15)
```
Great!! Exactly what I mean ❤️

bergjohan commented:
@SleeplessByte Wow, I wasn't expecting such a great review for this PR, thanks! :)
I've updated the solution based on your comments. I also switched to a more standard way of using references. Please let me know what you think! :)

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
```diff
@@ -0,0 +1,94 @@
+### Reasonable solutions
+
+Using a hand-rolled loop:
```
You might want to consider using heading at level 4 here. It's no requirement, but might look better in some views :)

```suggestion
#### Hand-rolled loop
```

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
````diff
@@ -0,0 +1,94 @@
+.loop_end:
+```
+
+Using the `rep movsb` instructions:
````
```suggestion
#### Using `rep movsb` instructions:
```

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
```diff
@@ -0,0 +1,94 @@
+
+### Common errors
+
+- Forgetting to save registers. "Registers `%rbp`, `%rbx` and `%r12` through `%r15` 'belong' to the calling function and the called function is required to preserve their values. In other words, a called function must preserve these registers’ values for its caller" (Matz, Hubička, Jaeger, & Mitchell, 2014, p. 15). Depending on which registers the caller is using, the tests might still pass.
```
```suggestion
- Forgetting to save registers. 

  > Registers `%rbp`, `%rbx` and `%r12` through `%r15` 'belong' to the calling function and the called function is required to preserve their values. In other words, a called function must preserve these registers’ values for its caller
  > — Matz, Hubička, Jaeger, & Mitchell, 2014, p. 15
  
  Depending on which registers the caller is using, the tests might still pass.
```

I think this was the intention? This is what it will look like:

### Common errors

- Forgetting to save registers. 

  > Registers `%rbp`, `%rbx` and `%r12` through `%r15` 'belong' to the calling function and the called function is required to preserve their values. In other words, a called function must preserve these registers’ values for its caller
  > — Matz, Hubička, Jaeger, & Mitchell, 2014, p. 15
  
  Depending on which registers the caller is using, the tests might still pass.

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
```diff
@@ -0,0 +1,94 @@
+
+- Forgetting to save registers. "Registers `%rbp`, `%rbx` and `%r12` through `%r15` 'belong' to the calling function and the called function is required to preserve their values. In other words, a called function must preserve these registers’ values for its caller" (Matz, Hubička, Jaeger, & Mitchell, 2014, p. 15). Depending on which registers the caller is using, the tests might still pass.
+
+- "Forgetting return. A function declaration must end with . . . `RET`. . . . The execution will continue in the code after the procedure if there is no `RET`" (Fog, 2019, p. 10).
```
```suggestion
- Forgetting return. 
  > A function declaration must end with . . . `RET`. . . . The execution will continue in the code after the procedure if there is no `RET`
  > — Fog, 2019, p. 10
```

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
```diff
@@ -0,0 +1,94 @@
+
+- If they are using the data segment (`.data`) for constants, suggest to use the read-only data segment (`.rodata`).
+
+- Prefer `lea rax, [rel symbol]` over `mov rax, symbol`. "There is absolutely no reason to use absolute addresses for simple memory operands. Rip-relative addresses make instructions shorter, they eliminate the need for relocation at load time, and they are safe to use in all systems" (Fog, 2019, p. 21).
```
```suggestion
- Prefer `lea rax, [rel symbol]` over `mov rax, symbol`. 
  > There is absolutely no reason to use absolute addresses for simple memory operands. Rip-relative addresses make instructions shorter, they eliminate the need for relocation at load time, and they are safe to use in all systems
  > — Fog, 2019, p. 21
```

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
```diff
@@ -0,0 +1,94 @@
+
+- Prefer `lea rax, [rel symbol]` over `mov rax, symbol`. "There is absolutely no reason to use absolute addresses for simple memory operands. Rip-relative addresses make instructions shorter, they eliminate the need for relocation at load time, and they are safe to use in all systems" (Fog, 2019, p. 21).
+
+- Clearing the direction flag using `cld` is redundant. "The direction flag `DF` in the `%rFLAGS` register must be clear (set to “forward” direction) on function entry and return" (Matz, Hubička, Jaeger, & Mitchell, 2014, p. 15)
```
```suggestion
- Clearing the direction flag using `cld` is redundant. 
  > The direction flag `DF` in the `%rFLAGS` register must be clear (set to “forward” direction) on function entry and return.
  > — Matz, Hubička, Jaeger, & Mitchell, 2014, p. 15
```

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
````diff
@@ -0,0 +1,94 @@
+
+- Minimize loop overhead. Example:
+
+    ```nasm
````
Looks like this needs less indentation.

````
- something
- something else
  ```text
  hello
  ```
````

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
````diff
@@ -0,0 +1,94 @@
+    .loop_end:
+    ```
+
+    "The most important problem with the loop . . . is that there are two jump instructions. We can eliminate one jump from the loop by putting the branch instruction in the end" (Fog, 2019, p. 89):
````
```suggestion
    > The most important problem with the loop . . . is that there are two jump instructions. We can eliminate one jump from the loop by putting the branch instruction in the end
    > — Fog, 2019, p. 89
```

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
```diff
@@ -0,0 +1,94 @@
+
+### Optimizations
+
+- Prefer `movzx edx, byte [rdi]` over `mov dl, byte [rdi]`. "Prevent false dependences by writing to a full register rather than a partial register" (Fog, 2019, p. 62).
```
```suggestion
- Prefer `movzx edx, byte [rdi]` over `mov dl, byte [rdi]`. 
  > Prevent false dependencies by writing to a full register rather than a partial register
  > — Fog, 2019, p. 62
```

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
```diff
@@ -0,0 +1,94 @@
+
+- Prefer `movzx edx, byte [rdi]` over `mov dl, byte [rdi]`. "Prevent false dependences by writing to a full register rather than a partial register" (Fog, 2019, p. 62).
+
+- Prefer `add/sub` over `inc/dec`. "The `INC` and `DEC` instructions are inefficient on some CPUs because they write to only part of the flags register (excluding the carry flag). Use `ADD` or `SUB` instead to avoid false dependences or inefficient splitting of the flags register" (Fog, 2019, p. 62).
```
```suggestion
- Prefer `add/sub` over `inc/dec`. 
  > The `INC` and `DEC` instructions are inefficient on some CPUs because they write to only part of the flags register (excluding the carry flag). Use `ADD` or `SUB` instead to avoid false dependencies or inefficient splitting of the flags register
  > — Fog, 2019, p. 62
```

SleeplessByte commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
```diff
@@ -0,0 +1,94 @@
+
+- Prefer `test dl, dl` over `cmp dl, 0` to reduce the size of the code.
+
+- Make conditional jumps most often not taken. "The efficiency and throughput for not-taken branches is better than for taken branches on most processors. Therefore, it is good to place the most frequent branch first" (Fog, 2019, p. 68).
```
```suggestion
- Make conditional jumps most often not taken. 
  > The efficiency and throughput for not-taken branches is better than for taken branches on most processors. Therefore, it is good to place the most frequent branch first
  > — Fog, 2019, p. 68
```

bergjohan commented:
@SleeplessByte Thanks for the review, I've updated the notes based on your comments! :)

kotp commented on tracks/x86-64-assembly/exercises/two-fer/mentoring.md:
```diff
@@ -0,0 +1,110 @@
+
+- Forgetting to save registers.
+
+    > Registers `%rbp`, `%rbx` and `%r12` through `%r15` “belong” to the calling function and the called function is required to preserve their values. In other words, a called function must preserve these registers’ values for its caller. (Matz, Hubička, Jaeger, & Mitchell, 2014, p. 15)
```
I am not sure how to get to page 15 of that document... but I did find "AMD64 ABI Draft 0.99 – December 7, 2007" on page 15.

kotp replied:
Found the footnote, just hadn't gotten that far down the page yet. ;)

kotp replied:
I hit approve, but also look forward to seeing the suggested formatting changes come in as well.

bergjohan replied:
Did I miss something? I thought I made all the suggested changes. Which formatting changes are we talking about?

kotp replied:
Perhaps the conversations just weren't marked as resolved.

kotp replied:
https://github.com/exercism/website-copy/pull/1352#discussion_r336793646 is an example.

SleeplessByte replied:
Yeah they show as outdated, just not resolved!

bergjohan commented:
Thanks everyone! :)
